### ISSUE TITLE
Admob Crash in Production Apps :com.google.android.gms.dynamite_dynamiteloader@17455081@17.4.55 (100400-248795830):4)

### ISSUE THREAD

**Issue by sotano42** [Scirra/Construct-3-bugs#3042]
In some devices (about the 3%) a Crash has been reported on my firebase console causing the App to close when requesting Interstitial videos from AdMob Network.

## Attach a .c3p

I haven't been able to reproduce the error, unfortunately I don't have all the types of the devices, here are the ones that crashed:
<li>
**Device**
Brand: samsung
<b>Model: Galaxy Note8</b>
Orientation: Portrait
RAM free: 1.9 GB
Disk free: 5.07 GB
**Operating System**
Version: 9
Orientation: Portrait
Rooted: No
**Crash**
Date: 14 Jun 2019, 17:28:00
</li>
<li>
**Device**
Brand: motorola
<b>Model: moto g(6) plus</b>
Orientation: Portrait
RAM free: 1.53 GB
Disk free: 35.47 GB
**Operating System**
Version: 9
Orientation: Portrait
Rooted: No
Crash
Date: 9 Jun 2019, 12:17:00
</li>
<li>
**Device**
Brand: motorola
<b>Model: moto g(6) plus</b>
Orientation: Portrait
RAM free: 1.49 GB
Disk free: 35.47 GB
**Operating System**
Version: 9
Orientation: Portrait
Rooted: No
**Crash**
Date: 9 Jun 2019, 12:16:00
</li>
<li>
**Device**
Brand: HUAWEI
<b>Model: Mate 20</b>
Orientation: Portrait
RAM free: 721.87 MB
Disk free: 46.46 GB
**Operating System**
Version: 9
Orientation: Portrait
Rooted: No
**Crash**
Date: 31 May 2019, 17:29:00
</li>

The error has presented when requesting an interstitial advert, haven't been able to reproduce it. I hope the stack. App immediately closes when the error occurs.

App shouldn't close.

App should just skip playing the ad if none is available.

## More details
I have to remark that the App was running without problems, and I suspect that AdMob may have changed something in their services that now causes the App to crash and that it is not handled propperly in the official AdMob cordova plugin.

I also have compiled my app using cordova CLI, and I'm using third party plugins. Altought I don't think they are related to the error, I'm sharing the list:

https://www.npmjs.com/package/cordova-plugin-firebase
https://www.npmjs.com/package/cordova-plugin-enable-multidex

**Affected browsers/platforms:** 
Android

**First affected release:** 
This started to happen on r148 (stable),

## System details

<details><summary>View details</summary>
<table>
<tr><td >
<b>Fatal Exception</b>: java.lang.NoClassDefFoundError: Failed resolution of: Lorg/apache/http/ProtocolVersion;
       at ly.b(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:2)
       at lx.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:1)
       at lz.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:29)
       at com.google.android.gms.ads.internal.util.aq.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:10)
       at lc.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:8)
       at lc.run(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:2)
</td></tr>
<tr><td>
<b>More info found on my firebase console regarding the class that failed <font color="red">:com.google.android.gms.dynamite_dynamiteloader</font></b>
</td></tr>
<tr><td>
<b>Caused by java.lang.ClassNotFoundException</b>: Didn't find class "org.apache.http.ProtocolVersion" on path: DexPathList[[zip file "/data/user_de/0/com.google.android.gms/app_chimera/m/000000c9/dl-AdsFdrDynamite.integ_20290000000.apk"],nativeLibraryDirectories=[/data/user_de/0/com.google.android.gms/app_chimera/m/000000c9/n/arm64-v8a, /system/lib64, /system/vendor/lib64]]
       at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
       at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
       at ab.loadClass(:com.google.android.gms.dynamite_dynamiteloader@17455081@17.4.55 (100400-248795830):4)
       at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
       at ly.b(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:2)
       at lx.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:1)
       at lz.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:29)
       at com.google.android.gms.ads.internal.util.aq.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:10)
       at lc.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:8)
       at lc.run(:com.google.android.gms.policy_ads_fdr_dynamite@20290000@20290000.249178941.249178941:2)
</td></tr>
<tr><td><b>The following dump presented with a different stack trace but may be related with previous, I'm pasting it here just in case it helps.</b><td></tr>
<tr><td><b>Fatal Exception:</b> java.util.concurrent.TimeoutException: com.google.android.gms.ads.internal.webview.v.finalize() timed out after 10 seconds
       at java.lang.Object.wait(Object.java)
       at com.google.android.gms.ads.exoplayer1.j.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290002@20290002.249178941.249178941:12)
       at com.google.android.gms.ads.exoplayer1.h.d(:com.google.android.gms.policy_ads_fdr_dynamite@20290002@20290002.249178941.249178941)
       at com.google.android.gms.ads.internal.video.exoplayer1.a.a(:com.google.android.gms.policy_ads_fdr_dynamite@20290002@20290002.249178941.249178941:1)
       at com.google.android.gms.ads.internal.video.gmsg.exoplayer1.f.b(:com.google.android.gms.policy_ads_fdr_dynamite@20290002@20290002.249178941.249178941:2)
       at com.google.android.gms.ads.internal.webview.v.Z(:com.google.android.gms.policy_ads_fdr_dynamite@20290002@20290002.249178941.249178941:1)
       at com.google.android.gms.ads.internal.webview.v.finalize(:com.google.android.gms.policy_ads_fdr_dynamite@20290002@20290002.249178941.249178941:4)
       at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:209)
       at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:192)
       at java.lang.Thread.run(Thread.java:818)</td></tr>
</td></tr>
</table> 
</details>

**Comment by AshleyScirra** [comment]
I've never heard of `com.google.android.gms.dynamite_dynamiteloader`. @shortercode, is this used by any of our official code? If not it would appear to be a crash in a third-party addon/library, which you'd need to take up with them instead.

**Comment by shortercode** [comment]
Given it's package ID it could be a dependency of the Admob SDK, but I'm not familiar with it's purpose. I did a bit of Googling and apparently there was a regression in Android 9 related to the same component with the Maps SDK. There's a workaround, but the issue was apparently resolved in services 16.1+. Our admob is currently locked to 16.0.0 due an incompatibility with 17.x , it's been on my TODO list for awhile to resolve that but we didn't have an urgent reason to do it.

I'll work on getting support for the latest version of admob working, hopefully that will resolve the issue. But it's sounds like it will take awhile for us to get feedback about if it makes any difference.

**Comment by sotano42** [comment]
Thanks guys. I guess there's not many people using crashlytics and may haven't noticed the problem, but I will try to keep you informed of any change in the crash trends as soon as the patch is uo and running.

El lun., 17 de junio de 2019 6:58 a. m., Iain Shorter <notifications@github.com<mailto:notifications@github.com>> escribió:

Given it's package ID it could be a dependency of the Admob SDK, but I'm not familiar with it's purpose. I did a bit of Googling and apparently there was a regression in Android 9 related to the same component with the Maps SDK. There's a workaround, but the issue was apparently resolved in services 16.1+. Our admob is currently locked to 16.0.0 due an incompatibility with 17.x , it's been on my TODO list for awhile to resolve that but we didn't have an urgent reason to do it.

I'll work on getting support for the latest version of admob working, hopefully that will resolve the issue. But it's sounds like it will take awhile for us to get feedback about if it makes any difference.

**Comment by QuasarX** [comment]
@sotano42 Hi, what plugin are you using for crashlytics with Construct 3?

**Comment by Boulerzzz** [comment]
I get hundreds of these as well. Started on June 4th. I can also keep you informed on any changes you make! I don't use any third party plugins.

**Comment by sotano42** [comment]
> @sotano42 Hi, what plugin are you using for crashlytics with Construct 3?

I wrote my own plugin implementation using firebase javascript API.

**Comment by shortercode** [comment]
Okay it took a little while to resolve but I've been able to update the plugin to work with the latest admob version. Due to a change in requirements I've had to add a cordova variable, which is linked to the Client ID (Android) property on the Admob object in the next release.

**Comment by sotano42** [comment]
> Okay it took a little while to resolve but I've been able to update the plugin to work with the latest admob version. Due to a change in requirements I've had to add a cordova variable, which is linked to the Client ID (Android) property on the Admob object in the next release.

Awesome news... will wait for the release and keep you notified about the crash trends. Thanks Ian.

**Comment by AshleyScirra** [comment]
r156 is now out, please test again!

**Comment by sotano42** [comment]
> r156 is now out, please test again!

Awesome... I will release an update of my game the next week. Will keep you posted if I see changes in the crash reports.

**Comment by sotano42** [comment]
Hi guys, to keep you updated: a couple of days ago I released a newer version of my game, trends seem to have a problem solved although I also noticed some changes in admob dumps, I believe AdMob team is performing some updates on their network, nevertheless crashes related with interstitial stood there.

I'll come back in a week or so, waiting to reach around 95% of version deploy on my players base to see if the problem got solved.

Trends:
![image](https://user-images.githubusercontent.com/10860045/60942046-c8e17380-a2a6-11e9-9d72-04723ccc9311.png)

![image](https://user-images.githubusercontent.com/10860045/60942056-d860bc80-a2a6-11e9-8646-7aa7b9f68ca5.png)

**Comment by sotano42** [comment]
@shortercode @AshleyScirra I've been monitoring the deployed version on PRD and I have not seen more crashes related with interstitials, although I've not reached the 100% of users it seems the last update to C3 got it fixed.

![image](https://user-images.githubusercontent.com/10860045/61571079-967a0800-aa56-11e9-987b-e07303e0f9f4.png)

And there's a clear trend of chrashes decreasing with the latest version of the game, so we may close this issue I guess.... thanks for your help. 

Trends:
![image](https://user-images.githubusercontent.com/10860045/61571111-c9bc9700-aa56-11e9-96c8-0dc8782b8085.png)

![image](https://user-images.githubusercontent.com/10860045/61571116-d6d98600-aa56-11e9-95fc-c5eeeaab989e.png)

![image](https://user-images.githubusercontent.com/10860045/61571159-333ca580-aa57-11e9-8867-be20cb5bc929.png)

And if by any mean you may need some help writting 3rd party plugins... I'll be around.

**Comment by QuasarX** [comment]
Why we don't have this plugin in Construct 3 :/

**Comment by sotano42** [comment]
This cordoca plugin has a google-services.json dependency so I guess that's the main reason, and you have to compile it using CLI, so you have to install NPM and Cordova and have a some coding skills, that could be another reason... I'm starting to think I should sell this... imos. 😅

**Comment by sotano42** [comment]
@QuasarX I released a couple of free plugins for C3, they may be useful for you.
[image](https://user-images.githubusercontent.com/10860045/61580087-6c1a6000-aad3-11e9-8f0c-0d40cdd3035a.png)

**Comment by QuasarX** [comment]
@sotano42 Thank you so much ✨✨✨

**Comment by sotano42** [comment]
> @sotano42 Thank you so much ✨✨✨
> Btw I can't see in plugin list and this link gives me "Permission Denied"
> http://www.imcsw.com/2019/07/15/handy-utilities-plugin/

That's weird... is public and I'm able to see it from several devices... try entering in the main webpage https://www.imcsw.com and find "Construct 3" at the top menu to navigate.

**Comment by QuasarX** [comment]
Ohh I mean Construct 3 plugin page give me "Permission Denied"
https://www.construct.net/en/make-games/addons/307/google-analytics-handy
![image](https://user-images.githubusercontent.com/13033837/61624326-11bf0200-ac81-11e9-9037-4baffeef53fb.png)

**Comment by shortercode** [comment]
Well it's good that it seems to be resolved. Part of me wonders if it was the update or if Admob was doing something bad for a short time... Doesn't matter at this point I guess.

Let me know if you have any further problems, and I'll take a look.

@QuasarX we did briefly look at firebase integration when the mobile advert plugin was under development, as Admob used to have quite tight integration with the firebase SDK. It was decided that it would be better to avoid needing to integrate firebase as well when it was "optional". There's still a case for needing firebase services in your app, while that required JSON file is a bit of pain for the build process it's probably doable. Feel free to file a feature request if one doesn't already exist.

@sotano42 I took a quick look at your plugin, it looks like it's not published ( I guess as an admin I can see unpublished plugins ).
